src/Shared/ChangeWaves.cs

```diff
@@ -0,0 +1,152 @@
+    {
+        Valid,
+        InvalidFormat,
+        OutOfRotation
+    }
+
+    /// <summary>
+    /// All waves are enabled by default, meaning all features behind change waves are enabled.
+    /// </summary>
+    internal class ChangeWaves
+    {
+        public static readonly string[] AllWaves = { Wave16_8, Wave16_10, Wave17_0 };
+        public const string Wave16_8 = "16.8";
+        public const string Wave16_10 = "16.10";
+        public const string Wave17_0 = "17.0";
+
+        /// <summary>
+        /// Special value indicating that all features behind change-waves should be enabled.
+        /// </summary>
+        public const string EnableAllFeatures = "999.999";
+
+        internal static readonly Version LowestWaveVersion = new Version(AllWaves[0]);
+        internal static readonly Version HighestWaveVersion = new Version(AllWaves[AllWaves.Length - 1]);
+        internal static readonly Version EnableAllFeaturesVersion = new Version(EnableAllFeatures);
+
+        internal static string LowestWave
+        {
+            get
+            {
+                return AllWaves[0];
+            }
+        }
+
+        internal static string HighestWave
+        {
+            get
+            {
+                return AllWaves[AllWaves.Length - 1];
+            }
+        }
+
+        private static string cachedWave = null;
+
+        internal static string DisabledWave
+        {
+            get
+            {
+                if (cachedWave == null)
+                {
+                    cachedWave = Traits.Instance.MSBuildDisableFeaturesFromVersion;
+                }
+
+                return cachedWave;
+            }
+            set
+            {
+                cachedWave = value;
+            }
+        }
+
+        /// <summary>
+        /// Ensure the the environment variable MSBuildDisableFeaturesFromWave is set to a proper value.
+        /// </summary>
+        /// <returns> String representation of the set change wave. "999.999" if unset or invalid, and the lowest version in rotation if out of bounds. </returns>
+        internal static string SanitizeChangeWave(out ChangeWaveConversionState result)
+        {
+            Version changeWave;
+
+            // If unset, enable all features.
+            if (string.IsNullOrEmpty(DisabledWave))
+            {
+                result = ChangeWaveConversionState.Valid;
+                return DisabledWave = ChangeWaves.EnableAllFeatures;
+            }
+
+            // If the user-set change wave is of invalid format, log a warning and enable all features.
+            if (!Version.TryParse(DisabledWave, out changeWave))
+            {
+                result = ChangeWaveConversionState.InvalidFormat;
+                return DisabledWave = ChangeWaves.EnableAllFeatures;
+            }
+
+            // If the change wave is out of rotation, log a warning and disable all features.
+            else if (ChangeWaves.IsChangeWaveOutOfRotation(changeWave))
```

Comment:
Being able to log the warning here would have been a lot easier, but it would need a logger of some sort to do that. It makes less sense that a function to let someone know if a change wave is enabled would require a logger to do so. 

So there was a bit of mental gymnastics to figure out a straightforward way to be able to log an error as we determine what the wave is, but I'm happy with the current implementation.